### PR TITLE
Add support for initial TDX VMCALL sub-functions

### PR DESCRIPTION
**Pull request by conradgrobler**
This PR adds Intel TDX support for:

- Sharing and unsharing guest-physical memory frames
- Reading from and writing to IO ports
- Reading from and writing to model-specific registers

### CODE REVIEW

File: oak_tdx_guest/src/vmcall.rs [hunk 1]
```diff
@@ -0,0 +1,528 @@
...
+    MapFailure(PhysAddr),
+}
+
+/// Maps a range of guest-physical addresses (GPAs) as shared with the hypervisor or guest-private.
+///
+/// If the shared bit of the GPA is set to 1 it will change the page mappings from private to
+/// shared.
+///
+/// If the guest changes pages back from shared to private, these will be added to the guest as
+/// pending pages, so the guest must also call TDCALL[TDG.MEM.PAGE.ACCEPT] before using them again.
```
Comment (tiziano88) [inline oak_tdx_guest/src/vmcall.rs:103]:
```suggestion
/// pending pages, so the guest must also call `TDCALL[TDG.MEM.PAGE.ACCEPT]` before using them again.
```

File: oak_tdx_guest/src/vmcall.rs [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+        .union(Registers::R13)
+        .union(Registers::R14);
+
+    let mut data: u64;
+
+    // The TDCALL leaf 0 goes into RAX. RAX returns the top-level result (0 is success). The
+    // bitflags of registers to be passed through to the VMM goes into RCX. The sub-function
+    // usage (always 0 when conforming to the GHCI spec) goes into R10, and the result of the
+    // subfunction is returned in R10. The sub-function to call goes into R11 and the data is
+    // returned in R11 if the read is successful. The size of the read (1,2 or 4 bytes) goest
```
Comment (tiziano88) [inline oak_tdx_guest/src/vmcall.rs:319]:
```suggestion
    // returned in R11 if the read is successful. The size of the read (1,2 or 4 bytes) goes
```

File: oak_tdx_guest/src/vmcall.rs [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+///
+/// See section 3.11 of [Guest-Host-Communication Interface (GHCI) for Intel® Trust Domain
+/// Extensions (Intel® TDX)](https://www.intel.com/content/dam/develop/external/us/en/documents/intel-tdx-guest-hypervisor-communication-interface.pdf)
+/// for more information.
+///
+/// #Safety
+///
+/// Modifying an MSR could impact the execution environment, so the caller must ensure that the MSR
+/// modification will not lead to undefined behavior.
+pub unsafe fn msr_write(msr: u32, data: u64) -> Result<(), &'static str> {
```
Comment (tiziano88) [inline oak_tdx_guest/src/vmcall.rs:429]: In general, I wonder if we should enable this: https://doc.rust-lang.org/beta/nightly-rustc/rustc_lint/builtin/static.UNSAFE_OP_IN_UNSAFE_FN.html

I find it confusing that `unsafe fn` implies that the entire body is itself an `unsafe { }` block.

WDYT?

**Review by tiziano88 — APPROVED** [review]

File: oak_tdx_guest/src/vmcall.rs (hunk 1 above)
Comment (conradgrobler) [reply oak_tdx_guest/src/vmcall.rs:103]: done

File: oak_tdx_guest/src/vmcall.rs (hunk 2 above)
Comment (conradgrobler) [reply oak_tdx_guest/src/vmcall.rs:319]: done

File: oak_tdx_guest/src/vmcall.rs (hunk 3 above)
Comment (conradgrobler) [reply oak_tdx_guest/src/vmcall.rs:429]: I think that is a good idea. I added it to this crate and create https://github.com/project-oak/oak/issues/3920 to track it more widely.

**Pull request merged**